File: src/agt_interfaces/test/test_localization_interfaces.cpp
#include <gtest/gtest.h>

#include <agt_interfaces/action/relocalize.hpp>
#include <agt_interfaces/msg/localization_status.hpp>

TEST(LocalizationStatusInterface, ConstantsAndFieldsAreUsable)
{
  agt_interfaces::msg::LocalizationStatus status;
  status.state = agt_interfaces::msg::LocalizationStatus::STATE_SEARCHING;
  status.error_code = agt_interfaces::msg::LocalizationStatus::ERROR_NONE;
  status.pose_valid = false;
  status.status_stale = false;
  status.message = "searching";

  EXPECT_EQ(status.state, 1U);
  EXPECT_EQ(status.error_code, 0U);
  EXPECT_FALSE(status.pose_valid);
  EXPECT_EQ(status.message, "searching");
}

TEST(RelocalizeInterface, GeneratedGoalFeedbackAndResultAreUsable)
{
  agt_interfaces::action::Relocalize::Goal goal;
  goal.mode = agt_interfaces::action::Relocalize::Goal::MODE_AUTO_SEARCH;
  goal.max_candidates = 8U;
  goal.timeout_s = 10.0;

  agt_interfaces::action::Relocalize::Feedback feedback;
  feedback.state = agt_interfaces::msg::LocalizationStatus::STATE_VERIFYING;
  feedback.tested_candidates = 2U;

  agt_interfaces::action::Relocalize::Result result;
  result.success = true;
  result.error_code = agt_interfaces::msg::LocalizationStatus::ERROR_NONE;
  result.final_status.state = agt_interfaces::msg::LocalizationStatus::STATE_TRACKING;

  EXPECT_EQ(goal.mode, 0U);
  EXPECT_EQ(goal.max_candidates, 8U);
  EXPECT_DOUBLE_EQ(goal.timeout_s, 10.0);
  EXPECT_EQ(feedback.state, 2U);
  EXPECT_EQ(feedback.tested_candidates, 2U);
  EXPECT_TRUE(result.success);
  EXPECT_EQ(result.final_status.state, 3U);
}

File: src/agt_interfaces/test/test_localization_interfaces.py
from agt_interfaces.action import Relocalize
from agt_interfaces.msg import LocalizationStatus


def test_localization_status_constants_and_defaults():
    status = LocalizationStatus()
    status.state = LocalizationStatus.STATE_SEARCHING
    status.error_code = LocalizationStatus.ERROR_AMBIGUOUS_RESULT
    status.status_stale = True

    assert status.state == 1
    assert status.error_code == 108
    assert LocalizationStatus.ERROR_STALE_SCAN == 113
    assert LocalizationStatus.ERROR_INVALID_SCAN_TIMESTAMP == 114
    assert status.status_stale is True
    assert status.pose_valid is False


def test_relocalize_generated_types_and_mode_constants():
    goal = Relocalize.Goal()
    goal.mode = Relocalize.Goal.MODE_SINGLE_INITIAL_POSE
    goal.use_initial_pose = True
    goal.max_candidates = 3

    feedback = Relocalize.Feedback()
    feedback.state = LocalizationStatus.STATE_TRACKING

    result = Relocalize.Result()
    result.success = False
    result.error_code = LocalizationStatus.ERROR_TIMEOUT

    assert goal.mode == 1
    assert goal.use_initial_pose is True
    assert goal.max_candidates == 3
    assert feedback.state == 3
    assert result.error_code == 106
